### ISSUE TITLE
DSP memory holds FIR coefficients in reverse order

### ISSUE THREAD

**Issue by mike811** [freeDSP/freeDSP-aurora#105]
Hello,

tests reveal a maximum phase behaviour of current FIR implementation.
This seems to be caused by incorrect order of coefficient upload. According to this:
https://wiki.analog.com/resources/tools-software/sigmastudio/toolbox/filters/firfilter
Coefficents need to be loaded in reverse order.

As a quick workaround coefficients may be swapped prior upload by a script or via command line e.g:
tail -r LIN.txt > LIN_rev.txt

**Comment by dspverden** [comment]
HI,

please can you verify wether the issue is solved in v2.2.2? I have added the coefficient order reversing to the ESP32 firmware.

**Comment by mike811** [comment]
Hello,

seems to work correctly in principle with pre 2.2.2 but there is a difference in frequency response:
For example with 2.1.4 and manual coef reversal:
<img width="684" alt="Bildschirmfoto 2022-03-15 um 19 02 36" src="https://user-images.githubusercontent.com/55871370/158442823-0a09d88e-8978-45aa-8fa4-cdc2fbdf0947.png">


With pre 2.2.2 and without manual coef reversal:
<img width="701" alt="Bildschirmfoto 2022-03-15 um 19 05 02" src="https://user-images.githubusercontent.com/55871370/158443018-f7ac2f59-047a-453b-a218-9c9332ac27d2.png">

Did something regarding PEQ change to cause this?

Regards
Michael

**Comment by dspverden** [comment]
Are there PEQs involved in your setting?
Or do I see the FIR only there?
I did test it with only FIR active and directly from audio interface back into audio interface and saw the expected FIR response.

**Comment by dspverden** [comment]
P.S. between 100 and 200 Hz I see a phase change, too. Can it be that you have enabled another filter (PEQ or Phase)?

**Comment by mike811** [comment]
No Aurora PEQ were involved. All 10 PEQs had Gain 0. Tested with clean 2.2.1 and pre 2.2.2 FW. Only FIR-inside EQs generated via rePhase.
Didn't you mention something about extending PEQs to 10 in FIRs plugin?

**Comment by dspverden** [comment]
Is there a way to show how the FIR should look like? So what did you design and does Aurora reproduce that correctly.
Or send me your filter and I can verify here.
Regarding PEQs: Yes the GUI was telling 10 PEQs but SigmaStudio has set it to 8 when copying the block from 8channels to 4FIRs.

**Comment by mike811** [comment]
Yes, the first picture with 2.1.4 shows my Woofer with linear phase close to zero. This was only tuned by rePhase EQs and should look like that ;-)

**Comment by dspverden** [comment]
ok, plz can you send me the text file you uploaded to aurora? I want to make sure that this is processed correctly.

**Comment by mike811** [comment]
You got mail.

**Comment by dspverden** [comment]
Yes saw it while I was typing here? Can I conclude from your previous post, that this issue is solved now?
Will have a look at your file anyway. Just to be sure with some files from outside of my lab.
Thanks for helping with debugging it.

**Comment by MarsianCRaute** [comment]
Great! Can't wait to use FIR more often now

**Comment by mike811** [comment]
Hello,
it seems there is still something going on with the FIR or EQ implementation.
After power up today I got this response measured:
<img width="1402" alt="Bildschirmfoto 2022-03-17 um 16 59 54" src="https://user-images.githubusercontent.com/55871370/158842675-1e94f09a-142d-4475-b38c-c1902c6e24db.png">

Only after reloading the coef file again I got the nice response from two days ago...

**Comment by dspverden** [comment]
I will verify this today evening.

**Comment by dspverden** [comment]
Hm, I cannot reproduce that here. Could it be that you did not press "Store Preset"?
Because in the IR file you have sent me, I see that you equalize the dip at 100. So the first picture looks a bit like no filter present at all. Btw. the file you have sent me does not fully match your result because I cannot see the LP behaviour in your measurement.
This is the file you have sent me:
![test1](https://user-images.githubusercontent.com/18739553/158886821-39ec2ce9-6ab6-4425-8ce2-6b7c0b76688f.jpg)
And this is what I measure after a reboot:
![test2](https://user-images.githubusercontent.com/18739553/158886981-f4bb3419-d5af-4d87-aaa8-e260bbbb1403.jpg)

**Comment by mike811** [comment]
Hi,
sorry for the confusion. I can't reproduce it here ether. You are probably right. Maybe I simply forgot to klick store preset when playing around last time...
